{
 "cells": [
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "# Panorama: A robust pangenome-based method for predicting and comparing biological systems across species\n",
    "\n",
    "Jérôme Arnoux$^{1,*}$, Jean Mainguy$^{1}$, Laura Bry$^{1}$, Quentin Fernandez de Grado$^{1,2}$, Vallenet David$^{1}$, Alexandra Calteau$^{1}$,\n",
    "\n",
    "\n",
    "1. LABGeM, Genomics Metabolics, CEA, Genoscope, Institut François Jacob, Université d’Évry, Université Paris-Saclay, CNRS\n",
    "2. Bla\n",
    "\n",
    "*Corresponding author*:\n",
    "* jarnoux@genoscope.cns.fr"
   ],
   "id": "e2c38aa7d5d91290"
  },
  {
   "cell_type": "markdown",
   "id": "2b5c99ec532cb5b1",
   "metadata": {},
   "source": [
    "# Abstrasct\n",
    "\n",
    "Over the last decade, the expansion in the number of available genomes has profoundly transformed the study of genetic diversity, evolution, and ecological adaptation in prokaryotes. However, traditional bioinformatic approaches based on the analysis of individual genomes are showing their limitations when faced with the sheer scale of the data. To overcome these constraints, the concept of pangenome has emerged, offering a comprehensive framework to capture the full genetic repertoire of a species. In this study, we present PANORAMA, an innovative pangenomic tool designed to exploit pangenome graphs and enable them to be annotated and compared in order to explore the genomic diversity of several species. Based on the PPanGGOLiN software suite, PANORAMA integrates advanced methods for rule-based prediction of macromolecular systems and comparative analysis of conserved features between different pangenomes, such as hotspots of insertion. We illustrate the use of PANORAMA on a Pseudomonas aeruginosa dataset, evaluating its performance against reference defense system prediction tools such as PADLOC and DefenseFinder. The analysis was then extended to a larger set including four species of Enterobacteriaceae ($>$6,000 genomes), demonstrating PANORAMA's ability to annotate, compare and explore the diversity and distribution of biological systems across multiple species. This work provides new methods for the large-scale comparative study of microbial genomes and underlines the relevance of pangenome approaches in deciphering their evolutionary dynamics. PANORAMA is freely available and accessible through: \\url{https://github.com/labgem/PANORAMA}\n",
    "\n",
    "**Keywords:**\n",
    "\n",
    "Pangenome, Comparative genomics, Bacterial defense systems, Pseudomonas aeruginosa, Enterobacteriaceae, Bioinformatics."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f4127bb36c78b50",
   "metadata": {},
   "source": [
    "# Introduction\n",
    "\n",
    "This Jupyter notebook will be organized into three parts. The first part (Results) will generate the raw figures presented in the article. The second (Supplementary) part will generate the additional figures. The last part (DATA) will contain the commands used to create the dataset, the pangenomes, and to redo the analyses with PANORAMA. These data are also available directly in the Zenodo repository linked to the publication. It is therefore not necessary to reproduce the data."
   ]
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "# Results\n",
    "\n",
    "## Define constant and load data\n"
   ],
   "id": "886509f75e2498bf"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-13T20:36:07.658787Z",
     "start_time": "2025-11-13T20:36:07.649246Z"
    }
   },
   "cell_type": "code",
   "source": [
    "from pathlib import Path\n",
    "from multiprocessing import cpu_count\n",
    "\n",
    "# Determine the number of available CPUs for multiprocessing\n",
    "CPU_COUNT = cpu_count()\n",
    "DISABLE_BAR = True\n",
    "DATA = Path(\"DATA/\")\n",
    "P_AERUGINOSA = DATA / \"P.aeruginosa/\"\n",
    "PANORAMA_PADLOC_RES = P_AERUGINOSA / \"PANORAMA/padloc/\"\n",
    "PADLOC_RES = P_AERUGINOSA / \"PADLOC/\"\n",
    "PANORAMA_DFINDER_RES = P_AERUGINOSA / \"PANORAMA/dfinder/\"\n",
    "DFINDER_RES = P_AERUGINOSA / \"DFINDER/\"\n",
    "GF2GENE = P_AERUGINOSA / \"gene_families.tsv\"\n",
    "LOCUS2PROTID = P_AERUGINOSA / \"locus2protID.tsv\"\n",
    "SYSTEM_FILE = \"systems.tsv\"\n",
    "OUTPUT = Path(\"figures\")\n",
    "OUTPUT.mkdir(parents=True, exist_ok=True)"
   ],
   "id": "a6b6c1d03d51229e",
   "outputs": [],
   "execution_count": 1
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-13T20:36:25.286721Z",
     "start_time": "2025-11-13T20:36:08.168730Z"
    }
   },
   "cell_type": "code",
   "source": [
    "from multiprocessing import Process, Manager\n",
    "from sources.read_results import load_gene_families, load_locus_tags\n",
    "\n",
    "# Create a manager to coordinate shared data\n",
    "manager = Manager()\n",
    "return_dict = manager.dict()\n",
    "\n",
    "# Starting the gene family and locus tag loading processes\n",
    "p1 = Process(target=load_gene_families, args=(GF2GENE, return_dict, DISABLE_BAR))\n",
    "p2 = Process(target=load_locus_tags, args=(LOCUS2PROTID, return_dict))\n",
    "\n",
    "# Start and join the first two processes to ensure dependent tasks can proceed\n",
    "p1.start()\n",
    "p2.start()\n",
    "p1.join()\n",
    "p2.join()\n",
    "\n",
    "# Retrieve necessary data from the return dictionary\n",
    "gene2family = return_dict['gene2family']\n",
    "locus2protID = return_dict['locus2protID']"
   ],
   "id": "4a8950b6746cb3d8",
   "outputs": [],
   "execution_count": 2
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-13T20:37:23.524390Z",
     "start_time": "2025-11-13T20:37:14.065151Z"
    }
   },
   "cell_type": "code",
   "source": [
    "from sources.dev_utils import get_tool_systems, merge_results\n",
    "from sources.read_results import load_padloc_results, load_dfinder_results, load_panorama_results, \\\n",
    "    read_panorama_pangenome_systems\n",
    "\n",
    "# Starting the result loading processes (dependent on gene2family and locus2protID)\n",
    "p3 = Process(target=load_padloc_results, args=(PADLOC_RES, gene2family, locus2protID, return_dict, DISABLE_BAR, 0))\n",
    "p4 = Process(target=load_panorama_results, args=(PANORAMA_PADLOC_RES / \"projection\", return_dict, \"padloc\", DISABLE_BAR, 1))\n",
    "p5 = Process(target=load_dfinder_results, args=(DFINDER_RES, gene2family, locus2protID, return_dict, DISABLE_BAR, 2))\n",
    "p6 = Process(target=load_panorama_results, args=(PANORAMA_DFINDER_RES / \"projection\", return_dict, \"dfinder\", DISABLE_BAR, 3))\n",
    "\n",
    "# Start the remaining processes\n",
    "p3.start()\n",
    "p4.start()\n",
    "p5.start()\n",
    "p6.start()\n",
    "\n",
    "# Wait for all processes to complete\n",
    "p3.join()\n",
    "p4.join()\n",
    "p5.join()\n",
    "p6.join()\n",
    "\n",
    "# Retrieving results from the shared dictionary\n",
    "fam2size = return_dict['fam2size']\n",
    "\n",
    "# Aggregating counts and annotations\n",
    "count_dict = {\n",
    "    \"Padloc\": return_dict['padloc_count'],\n",
    "    \"Panorama_padloc\": return_dict['panorama_padloc_count'],\n",
    "    \"Dfinder\": return_dict['dfinder_count'],\n",
    "    \"Panorama_dfinder\": return_dict['panorama_dfinder_count']\n",
    "}\n",
    "\n",
    "fam2annot = {\n",
    "    \"Padloc\": return_dict['padloc_fam_annot'],\n",
    "    \"Panorama_padloc\": return_dict['panorama_padloc_fam_annot'],\n",
    "    \"Dfinder\": return_dict['dfinder_fam_annot'],\n",
    "    \"Panorama_dfinder\": return_dict['panorama_dfinder_fam_annot']\n",
    "}\n",
    "\n",
    "systems2partitions = {\n",
    "    \"Padloc\": return_dict['padloc_systems2partitions'],\n",
    "    \"Dfinder\": return_dict['dfinder_systems2partitions']\n",
    "}\n",
    "\n",
    "name2systems = {\n",
    "    \"Padloc\": return_dict['padloc_name2systems'],\n",
    "    \"Dfinder\": return_dict['dfinder_name2systems']\n",
    "}\n",
    "\n",
    "# Retrieve results for each method\n",
    "padloc_res = return_dict['padloc_res']\n",
    "dfinder_res = return_dict['dfinder_res']\n",
    "panorama_padloc_res = return_dict['panorama_padloc_res']\n",
    "panorama_dfinder_res = return_dict['panorama_dfinder_res']\n",
    "\n",
    "# Extract systems from PADLOC and DefenseFinder\n",
    "padloc_systems = get_tool_systems(padloc_res)\n",
    "dfinder_systems = get_tool_systems(dfinder_res)\n",
    "\n",
    "# Extract systems from the panorama dictionaries\n",
    "panorama_systems_padloc = read_panorama_pangenome_systems(PANORAMA_PADLOC_RES / SYSTEM_FILE)\n",
    "panorama_systems_dfinder = read_panorama_pangenome_systems(PANORAMA_DFINDER_RES / SYSTEM_FILE)\n",
    "# Combine the panorama systems\n",
    "dfinder_systems = merge_results(dfinder_systems, dfinder_systems, 1)\n",
    "padloc_systems = merge_results(padloc_systems, padloc_systems, 1)\n",
    "panorama_systems = merge_results(panorama_systems_padloc, panorama_systems_dfinder, 1)"
   ],
   "id": "eb9fcc0cb3ecaa5f",
   "outputs": [],
   "execution_count": 3
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "## System prediction benchmark",
   "id": "8a31b4c9bc731b77"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "from IPython.display import display\n",
    "import pandas as pd\n",
    "from sources.boxplot import gen_boxplot_pangenome\n",
    "from sources.venn import plot_venn_diagram, count_system_similar\n",
    "from sources.score import compute_score_pangenome, compute_score_organisms\n",
    "\n",
    "P_AERUGINOSA_OUTPUT = OUTPUT / \"P.aeruginosa\"\n",
    "P_AERUGINOSA_OUTPUT.mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "gen_boxplot_pangenome(padloc=padloc_systems, dfinder=dfinder_systems,\n",
    "                      panorama=panorama_systems, output=P_AERUGINOSA_OUTPUT / \"boxplot_venn.png\")\n",
    "plot_venn_diagram(padloc_systems, dfinder_systems, panorama_systems, P_AERUGINOSA_OUTPUT)\n",
    "\n",
    "count_similar = count_system_similar(padloc_systems, dfinder_systems, panorama_systems_padloc,\n",
    "                                     panorama_systems_dfinder, P_AERUGINOSA_OUTPUT, 1)\n",
    "display(count_similar)\n",
    "THRESHOLD = 0.5\n",
    "scores = {\"Pangenome\": {\"PADLOC\": {},\n",
    "                        \"DFinder\": {}},\n",
    "          \"Organisms\": {\"PADLOC\": {},\n",
    "                        \"DFinder\": {}}\n",
    "          }\n",
    "recall, precision, f1_score = compute_score_pangenome(panorama_systems_padloc, padloc_systems, 'PADLOC',\n",
    "                                                      P_AERUGINOSA_OUTPUT, THRESHOLD)\n",
    "scores[\"Pangenome\"][\"PADLOC\"] = {\"Recall\": recall, \"Precision\": precision, \"F1-Score\": f1_score}\n",
    "\n",
    "recall, precision, f1_score = compute_score_pangenome(panorama_systems_dfinder, dfinder_systems, 'Defense_Finder',\n",
    "                                                      P_AERUGINOSA_OUTPUT, THRESHOLD)\n",
    "scores[\"Pangenome\"][\"DFinder\"] = {\"Recall\": recall, \"Precision\": precision, \"F1-Score\": f1_score}\n",
    "\n",
    "recall, precision, f1_score = compute_score_organisms(panorama_padloc_res, padloc_res, 'PADLOC', P_AERUGINOSA_OUTPUT,\n",
    "                                                      THRESHOLD)\n",
    "scores[\"Organisms\"][\"PADLOC\"] = {\"Recall\": recall, \"Precision\": precision, \"F1-Score\": f1_score}\n",
    "\n",
    "recall, precision, f1_score = compute_score_organisms(panorama_dfinder_res, dfinder_res, 'Defense_Finder',\n",
    "                                                      P_AERUGINOSA_OUTPUT,\n",
    "                                                      THRESHOLD)\n",
    "scores[\"Organisms\"][\"DFinder\"] = {\"Recall\": recall, \"Precision\": precision, \"F1-Score\": f1_score}\n",
    "\n",
    "# Convert scores dictionary to DataFrame\n",
    "scores_list = []\n",
    "for category, tools in scores.items():\n",
    "    for tool, metrics in tools.items():\n",
    "        scores_list.append({\n",
    "            \"Category\": category,\n",
    "            \"Tool\": tool,\n",
    "            \"Recall\": metrics[\"Recall\"],\n",
    "            \"Precision\": metrics[\"Precision\"],\n",
    "            \"F1-Score\": metrics[\"F1-Score\"]\n",
    "        })\n",
    "\n",
    "scores_df = pd.DataFrame(scores_list)\n",
    "display(scores_df)"
   ],
   "id": "5ac0e315ea5c5392",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    " **Fig. 2. Comparison of PANORAMA, PADLOC, and DefenseFinder system predictions at the pangenome\n",
    "level.** PADLOC and Defense Finder predictions at the genome level were unified at the pangenome level by converting sets\n",
    "of system genes to sets of gene families. (A) Venn diagram illustrating the overlap of system families predicted by the three\n",
    "tools. (B) Boxplots displaying the distribution of system counts for each category, as predicted by the different tools.\n",
    "\n",
    "## *Pseudomonas aeruginosa* defense system analysis\n",
    "\n",
    "### System prediction and analysis"
   ],
   "id": "d05a2d91e8cab669"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "import pandas as pd\n",
    "from sources.dev_utils import type2category\n",
    "from sources.system_analyses import plot_occurrences_and_fraction\n",
    "\n",
    "MODELS = DATA / \"Models/defense-finder/models_list.tsv\"\n",
    "RESULTS = PANORAMA_DFINDER_RES / 'systems.tsv'\n",
    "NB_GROUP = 20\n",
    "NB_COLS = 55\n",
    "NB_SPOTS = 7\n",
    "FIGSIZE = (40, 20)\n",
    "REVERSE = False\n",
    "BAR_COLORS = (\"darkblue\", \"darkgreen\", \"darkorange\")\n",
    "\n",
    "systems_data = pd.read_csv(RESULTS, sep='\\t',\n",
    "                           usecols=[\"system number\", \"system name\", \"organism\",\n",
    "                                    \"model_GF\", \"partition\", \"spots\"])\n",
    "\n",
    "# Add the system category grouping\n",
    "systems_data['system category'] = systems_data['system name'].map(type2category)\n",
    "systems_data.to_csv(OUTPUT / \"sys_tmp.tsv\", sep=\"\\t\", index=False)\n",
    "plot_occurrences_and_fraction(systems_data, MODELS, NB_GROUP, field='system category', output=P_AERUGINOSA_OUTPUT,\n",
    "                              kind=\"bar\", reverse=REVERSE, bar_colors=BAR_COLORS)"
   ],
   "id": "672fca24a80122b7",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "**Fig. 3. System prediction metrics in *P. aeruginosa*.** Systems are grouped by categories on the x-axis and ordered\n",
    "by decreasing values. Only the 19 highest-value system categories are displayed, with others grouped under the \"Other\"\n",
    "category. (A) Number of systems found for each category in the pangenome. (B) Relative frequency of system categories in\n",
    "genomes. (C) Shannon entropy of system categories.\n",
    "\n",
    "### Defense islands and spots of insertion"
   ],
   "id": "8a7d7cfab7aff0a0"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "from sources.hotspot import get_spot2orgs, get_data_projection, scatter_plot\n",
    "\n",
    "projection = PANORAMA_DFINDER_RES / \"projection\"\n",
    "tables = P_AERUGINOSA / 'table'\n",
    "\n",
    "number_genomes = sum(1 for _ in tables.glob(\"*.tsv\"))\n",
    "spot_to_orgs = get_spot2orgs(tables)\n",
    "proj_df = get_data_projection(projection)\n",
    "\n",
    "proj_df.to_csv(P_AERUGINOSA_OUTPUT / \"spot2sys_genomes.tsv\", sep=\"\\t\", index=True)\n",
    "scatter_plot(proj_df, spot_to_orgs, number_genomes, P_AERUGINOSA_OUTPUT, file_name=\"defense_hotspot_aeruginosa\",\n",
    "             min_freq=.25, min_sys=60)\n"
   ],
   "id": "6e170b0446bd75b8",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "**Fig. 4. System diversity and defense spot frequency in *P. aeruginosa*.** This bubble plot displays the distribution\n",
    "of defense spots identified by PANORAMA, based on their frequency in genomes (x-axis) and the total number of defense\n",
    "systems identified within each spot at the pangenome level (y-axis). The size of the bubbles is proportional to the number\n",
    "of distinct system categories represented in each spot (legend at the top shows scale)."
   ],
   "id": "f144dae6fb926822"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "from sources.system_analyses import plot_spots_and_pies\n",
    "\n",
    "plot_spots_and_pies(proj_df, systems_data, field='system category', nb_cols=NB_COLS,\n",
    "                    nb_spots=NB_SPOTS, output=P_AERUGINOSA_OUTPUT, figsize=FIGSIZE)"
   ],
   "id": "ad3adee394dfd610",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "**Fig. 5. Defense systems within insertion spots of the *P. aeruginosa* pangenome.** The bar plot (top) displays the\n",
    "number of predicted defense systems in the P. aeruginosa pangenome for each insertion spot. Only spots with at least 10\n",
    "systems are displayed. The pie charts (bottom) illustrate the system category composition for the six major insertion spots.\n",
    "\n",
    "## Pangenome comparison of Enterobacteriaceae defense arsenal\n",
    "\n",
    "### Get results and define constant"
   ],
   "id": "beaef584c4ea836f"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# Determine the number of available CPUs for multiprocessing\n",
    "ENTEROBACTERIACEAE = DATA / \"Enterobacteriaceae/\"\n",
    "ENTEROBACTERIACEAE_OUTPUT = OUTPUT / \"Enterobacteriaceae\"\n",
    "ENTEROBACTERIACEAE_OUTPUT.mkdir(parents=True, exist_ok=True)\n",
    "C_FREUNDII = ENTEROBACTERIACEAE / \"C.freundii/dfinder\"\n",
    "C_FREUNDII_OUTPUT = ENTEROBACTERIACEAE_OUTPUT / \"C.freundii\"\n",
    "C_FREUNDII_OUTPUT.mkdir(parents=True, exist_ok=True)\n",
    "E_COLI = ENTEROBACTERIACEAE / \"E.coli/dfinder\"\n",
    "E_COLI_OUTPUT = ENTEROBACTERIACEAE_OUTPUT / \"E.coli\"\n",
    "E_COLI_OUTPUT.mkdir(parents=True, exist_ok=True)\n",
    "K_PNEUMONIAE = ENTEROBACTERIACEAE / \"K.pneumoniae/dfinder\"\n",
    "K_PNEUMONIAE_OUTPUT = ENTEROBACTERIACEAE_OUTPUT / \"K.pneumoniae\"\n",
    "K_PNEUMONIAE_OUTPUT.mkdir(parents=True, exist_ok=True)\n",
    "S_ENTERICA = ENTEROBACTERIACEAE / \"S.enterica/dfinder\"\n",
    "S_ENTERICA_OUTPUT = ENTEROBACTERIACEAE_OUTPUT / \"S.enterica\"\n",
    "S_ENTERICA_OUTPUT.mkdir(parents=True, exist_ok=True)"
   ],
   "id": "9aa2d2f0ac61f5c",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "### Defense system distribution in the four species",
   "id": "49a85fb423eefd74"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "from sources.heatmap import load_system_data, generate_heatmap\n",
    "\n",
    "heatmap_data = load_system_data(ENTEROBACTERIACEAE)\n",
    "params = {\n",
    "    'linecolor': 'lightgray',\n",
    "    'linewidths': 1,\n",
    "    \"cmap\": 'rocket_r',\n",
    "    \"xticklabels\": True,\n",
    "    \"yticklabels\": True\n",
    "}\n",
    "\n",
    "# Generate a heatmap for normalized system frequencies without RM\n",
    "heatmap_data_normalized = heatmap_data.div(heatmap_data.sum(axis=0), axis=1) * 100\n",
    "heatmap_data_normalized_no_rm = heatmap_data_normalized.drop(index=\"RM\")\n",
    "generate_heatmap(\n",
    "    heatmap_data_normalized_no_rm,\n",
    "    ENTEROBACTERIACEAE_OUTPUT / \"heatmap_normalized_sys_no_RM.png\",\n",
    "    figsize=FIGSIZE,\n",
    "    cbar_label=\"System frequency\",\n",
    "    annot=True,\n",
    "    fmt=\".1f\",\n",
    "    rotation=90,\n",
    "    **params\n",
    ")\n"
   ],
   "id": "aeb994ade59fc59f",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "**Fig. 6. Relative frequency of system categories in Enterobacteriaceae pangenomes.** The relative frequencies are expressed as a percentage. The RM system category was removed to get a clearer view.\n",
    "\n",
    "### Identification of conserved spots"
   ],
   "id": "6f5d16229330d292"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "from sources.upset_plot import parse_data, generate_upset_plot_spots\n",
    "\n",
    "# Example usage\n",
    "cs_res = pd.read_csv(ENTEROBACTERIACEAE / \"all_conserved_spots.tsv\", sep=\"\\t\",\n",
    "                     usecols=[\"Conserved_Spot_ID\", \"Spot_ID\", \"Pangenome\"])\n",
    "\n",
    "spots_data = {\n",
    "    'C.freundii': 254,\n",
    "    'E.coli': 2036,\n",
    "    'K.pneumoniae': 778,\n",
    "    'S.enterica': 458\n",
    "}\n",
    "\n",
    "systems_data = {\n",
    "    'C.freundii': pd.read_csv(C_FREUNDII / SYSTEM_FILE, sep=\"\\t\"),\n",
    "    'E.coli': pd.read_csv(E_COLI / SYSTEM_FILE, sep=\"\\t\"),\n",
    "    'K.pneumoniae': pd.read_csv(K_PNEUMONIAE / SYSTEM_FILE, sep=\"\\t\"),\n",
    "    'S.enterica': pd.read_csv(S_ENTERICA / SYSTEM_FILE, sep=\"\\t\")\n",
    "}\n",
    "\n",
    "parsed_data = parse_data(cs_res, systems_data, spots_data)\n",
    "parsed_data.to_csv(ENTEROBACTERIACEAE_OUTPUT / \"upset_data.tsv\", sep=\"\\t\", index=False)\n",
    "generate_upset_plot_spots(parsed_data, list(systems_data.keys()), ENTEROBACTERIACEAE_OUTPUT)\n",
    "\n"
   ],
   "id": "e43d9484667b2f1a",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "**Fig. 7. Sharing of spot clusters across four species and their association with defense systems.** The UpSet plot shows the number of spot clusters shared across the four compared species, with stacked bars to indicate whether they contain defense systems (in blue) or not (in orange). The two top bar plots represent spot and defense system abundance metrics on a logarithmic scale.\n",
    "\n",
    "### Spot identification and conservation"
   ],
   "id": "88dbc7123db9e81e"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "from sources.cs_link import parse_data, generate_tile_plot_visualization\n",
    "\n",
    "results = pd.read_csv(ENTEROBACTERIACEAE / \"all_conserved_spots.tsv\", sep=\"\\t\",\n",
    "                      usecols=[\"Conserved_Spot_ID\", \"Spot_ID\", \"Pangenome\"])\n",
    "\n",
    "systems_data = {\n",
    "    'C.freundii': pd.read_csv(C_FREUNDII / SYSTEM_FILE, sep=\"\\t\"),\n",
    "    'E.coli': pd.read_csv(E_COLI / SYSTEM_FILE, sep=\"\\t\"),\n",
    "    'K.pneumoniae': pd.read_csv(K_PNEUMONIAE / SYSTEM_FILE, sep=\"\\t\"),\n",
    "    'S.enterica': pd.read_csv(S_ENTERICA / SYSTEM_FILE, sep=\"\\t\")\n",
    "}\n",
    "\n",
    "data = parse_data(results, systems=systems_data)\n",
    "data.to_csv(ENTEROBACTERIACEAE_OUTPUT / \"cs_link.tsv\", sep=\"\\t\", index=True)\n",
    "generate_tile_plot_visualization(data, ENTEROBACTERIACEAE_OUTPUT)\n"
   ],
   "id": "e04cb907a97e5538",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "**Fig. 8. Relative frequency of system categories in each spot cluster.** The heatmap presents the relative frequency\n",
    "(%) of system categories per spot cluster. Species associated with each spot cluster are listed next to the cluster identifiers."
   ],
   "id": "6d86f3383ffaf50e"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "# Supplementary\n",
   "id": "732957e15e2ce1ca"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "**Fig. S1. Number of Systems Predicted for Each Model in Enterobacteriaceae Pangenomes.** PANORAMA automatically generates this figure. PANORAMA employs the Bokeh package to create interactive and dynamic visualizations, which can be directly accessed and manipulated in any standard web browser.",
   "id": "e0bf5e441782da2"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# N.B: You must launch the cell that generates the figure 6\n",
    "\n",
    "# Calculate and plot system specificity\n",
    "col_sums = heatmap_data.sum(axis=0).to_numpy()  # Sum by species (columns)\n",
    "row_sums = heatmap_data.sum(axis=1).to_numpy()  # Sum by system (rows)\n",
    "total_sum = heatmap_data.values.sum()  # Total sum\n",
    "\n",
    "# Calculate specificity metric\n",
    "specificity_data = (heatmap_data.div(col_sums, axis=1)) / (row_sums / total_sum)[:, None]\n",
    "generate_heatmap(\n",
    "    specificity_data.T,\n",
    "    ENTEROBACTERIACEAE_OUTPUT / \"heatmap_system_distribution.png\",\n",
    "    figsize=(15, 35),\n",
    "    cbar_label=\"Specificity\",\n",
    "    annot=True,\n",
    "    fmt=\".1f\",\n",
    "    **params\n",
    ")"
   ],
   "id": "9af18640fd999533",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "**Fig. S2. Species-specificity evaluation across system categories in Enterobacteriaceae Pangenomes.** The\n",
    "enrichment factors were computed using the method described in Equation 6, providing a quantitative measure of species-specific representation."
   ],
   "id": "c7f5f28f5e996d8f"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from sources.dev_utils import type2category\n",
    "from sources.system_analyses import get_pie_charts_info, plot_pie, create_color_map, spot_occurrence, \\\n",
    "    plot_spot_occurrence, parse_data\n",
    "\n",
    "\n",
    "def plot_spot_systems(proj_data: pd.DataFrame, sys_data: pd.DataFrame, output: Path):\n",
    "    \"\"\"\n",
    "    Generate a plot with occurrences in spots and pie charts for the top spots.\n",
    "\n",
    "    Args:\n",
    "        sys_data (pd.DataFrame): DataFrame containing the system data.\n",
    "        output (Path): Path to save the output plots.\n",
    "\n",
    "    Returns:\n",
    "        None\n",
    "    \"\"\"\n",
    "    nb_slice = 10\n",
    "    nb_top_occ = 24\n",
    "    field = 'system category'\n",
    "    # Get spot occurrences and top types\n",
    "    data_df = parse_data(proj_data, sys_data)\n",
    "    spot_grouped_type_counts, top_types_occ = spot_occurrence(data_df, output, field=field, nb_cols=20,\n",
    "                                                              nb_top=nb_top_occ, keep_other_spots=False)\n",
    "\n",
    "    # Get pie chart information\n",
    "    top_types_pie, top_spots, spot_type_counts = get_pie_charts_info(data_df, field, 5, nb_slice)\n",
    "\n",
    "    # Create a consistent color map for system types\n",
    "    color_map = create_color_map(top_types_occ | top_types_pie)\n",
    "\n",
    "    # Plot and save bar plot for spot occurrences\n",
    "    nb_col_legend = len(top_types_occ) // 10 + 1\n",
    "    fig, ax = plt.subplots(figsize=FIGSIZE)\n",
    "    plot_spot_occurrence(spot_grouped_type_counts, color_map, nb_color_legend=nb_col_legend, ax=ax)\n",
    "    ax.set_title(\"Occurrences in spots\", fontsize=20)\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(output / \"spots_sys_distri.png\", bbox_inches=\"tight\", dpi=300)\n",
    "\n",
    "\n",
    "def draw_pie_chart(sys_data: pd.DataFrame, spot_id: int, output: Path, fig_name: str):\n",
    "    \"\"\"\n",
    "    Generates and saves a pie chart visualization based on the specified spot data. The method uses\n",
    "    the system data and a specific spot identifier to create color mappings and plot a pie chart\n",
    "    visualizing a breakdown of the spot type counts. The figure is saved in both PNG and SVG formats\n",
    "    with high resolution.\n",
    "\n",
    "    Args:\n",
    "        sys_data: A pandas DataFrame containing the system data to be aggregated and visualized.\n",
    "        spot_id: An integer representing the unique identifier of the spot for which the pie chart\n",
    "                is being generated.\n",
    "    Returns:\n",
    "         None\n",
    "    \"\"\"\n",
    "    top_types_pie, _, spot_type_counts = get_pie_charts_info(sys_data, field=\"system category\", nb_spots=6,\n",
    "                                                             nb_slice=10, top_spots=[str(spot_id)])\n",
    "    color_map = create_color_map(top_types_pie)\n",
    "    spot_data = spot_type_counts.loc[str(spot_id)].drop('total')\n",
    "    fig, ax = plt.subplots(figsize=(8, 12))\n",
    "    plot_pie(spot_data, color_map, ax, 10, base_fontsize=24)\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(output / f\"{fig_name}.png\", format='png', dpi=300)\n",
    "    plt.savefig(output / f\"{fig_name}.svg\", format='svg')\n"
   ],
   "id": "fe81575eede1ea37",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "projection = S_ENTERICA / \"projection\"\n",
    "proj_df = get_data_projection(projection)"
   ],
   "id": "219513fdc47ca961",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "s_enterica_systems = pd.read_csv(S_ENTERICA / SYSTEM_FILE, sep=\"\\t\")\n",
    "s_enterica_systems[\"system category\"] = s_enterica_systems[\"system name\"].map(type2category)\n",
    "plot_spot_systems(proj_df, s_enterica_systems, output=S_ENTERICA_OUTPUT)\n",
    "draw_pie_chart(s_enterica_systems, 86, S_ENTERICA_OUTPUT, \"spots_86_enterica\")"
   ],
   "id": "b992558846f5d720",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "projection = E_COLI / \"projection\"\n",
    "proj_df = get_data_projection(projection)"
   ],
   "id": "901ffa94ff13e9c1",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "e_coli_systems = pd.read_csv(E_COLI / SYSTEM_FILE, sep=\"\\t\")\n",
    "e_coli_systems[\"system category\"] = e_coli_systems[\"system name\"].map(type2category)\n",
    "plot_spot_systems(proj_df, e_coli_systems, output=E_COLI_OUTPUT)\n",
    "draw_pie_chart(e_coli_systems, 175, E_COLI_OUTPUT, \"spots_175_coli\")"
   ],
   "id": "97484eea0db7139c",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "# DATA\n",
    "\n",
    "## Where to find the information\n",
    "\n",
    "In the zenodo, you can find either the genomes' assembly summary for each species or the pangenome already computed. This way you can start from genomes or pangenomes.\n",
    "\n",
    "## Create Conda environments\n",
    "In this part we suggest creating new conda environments to prevent any incompatibility.\n",
    "\n",
    "### PANORAMA\n",
    "\n",
    "To install PPanGGOLiN & PANORAMA proceed as such:\n",
    "\n",
    "```shell\n",
    "# Create a new Conda environment and install PPanGGOLiN\n",
    "conda create -n panorama -c conda-forge -c bioconda panorama\n",
    "\n",
    "# Activate the environment\n",
    "conda activate panorama\n",
    "\n",
    "# Verify the installation\n",
    "ppanggolin --version\n",
    "panorama --version\n",
    "\n",
    "# Install also genome_updater to download genomes\n",
    "conda install -c bioconda genome_updater\n",
    "conda deactivate\n",
    "```\n",
    "\n",
    "### PADLOC\n",
    "\n",
    "Install PADLOC such as:\n",
    "```shell\n",
    "# Install PADLOC into a new conda environment\n",
    "conda create -n padloc -c conda-forge -c bioconda -c padlocbio padloc==2.0.0\n",
    "# Activate the environment\n",
    "conda activate padloc\n",
    "# Install hmmer\n",
    "conda install -c bioconda hmmer\n",
    "# Download the latest database\n",
    "padloc --db-update\n",
    "```\n",
    "\n",
    "### DefenseFinder\n",
    "\n",
    "Install DefenseFinder such as:\n",
    "```shell\n",
    "conda create -n defensefinder\n",
    "conda activate defensefinder\n",
    "pip install mdmparis-defense-finder==1.2.2\n",
    "conda install hmmer\n",
    "defense-finder update\n",
    "```\n",
    "\n",
    "## P.aeruginosa\n",
    "### Download genomes\n",
    "\n",
    "```shell\n",
    "conda activate panorama\n",
    "cd DATA/P.aeruginosa\n",
    "mkdir Genomes\n",
    "genome_updater.sh -M gtdb -e assembly_summary.txt -f 'genomic.fna.gz, assembly_report.txt, protein.faa.gz,  genomic.gff.gz, translated_cds.faa.gz' -o Genomes -t 20\n",
    "```\n",
    "### Generate the pangenome\n",
    "\n",
    "```shell\n",
    "cd DATA/P.aeruginosa\n",
    "mkdir Pangenome\n",
    "if [ -f \"Pangenome/P.aeruginosa_gff.list\" ]; then\n",
    "    rm \"Pangenome/P.aeruginosa_gff.list\"\n",
    "fi\n",
    "if [ -f \"Pangenome/P.aeruginosa_fna.list\" ]; then\n",
    "    rm \"Pangenome/P.aeruginosa_fna.list\"\n",
    "fi\n",
    "for gff in Genomes/*/files/*gff*; do fna=$(echo $gff | sed 's/gff/fna/'); org=$(echo $gff | cut -d'/' -f4 | sed 's/_genomic.gff.gz//'); echo $org    $(pwd)/$gff >>Pangenome/P.aeruginosa_gff.list; echo $org    $(pwd)/$fna >> Pangenome/P.aeruginosa_fna.list; done\n",
    "sed -i 's/\\s\\s*/\\t/' Pangenome/P.aeruginosa_*\n",
    "ppanggolin all --anno Pangenome/P.aeruginosa_gff.list --fasta Pangenome/P.aeruginosa_fna.list --no_flat_files --cpu 20 -o Pangenome -f\n",
    "```\n",
    "\n",
    "### Prediction with PADLOC\n",
    "\n",
    "First, we need to uncompress the genomes as PADLOC does not read compressed files.\n",
    "\n",
    "```shell\n",
    "conda activate padloc\n",
    "cd DATA/P.aeruginosa\n",
    "mkdir -p PADLOC/tmp_genomes/GFF\n",
    "for gff in Genomes/*/files/*gff*; do name=$(basename \"$gff\" .gz); gunzip -c $gff > PADLOC/tmp_genomes/GFF/$name; done\n",
    "mkdir -p PADLOC/tmp_genomes/FAA\n",
    "for faa in Genomes/*/files/*_protein.faa*; do name=$(basename \"$faa\" .gz); gunzip -c $faa > PADLOC/tmp_genomes/FAA/$name; done\n",
    "cd ..\n",
    "```\n",
    "\n",
    "PADLOC introduces \"other\" models that are not corresponding to an existing model. In our analyses we removed this model.\n",
    "\n",
    "```shell\n",
    "git clone https://github.com/padlocbio/padloc-db.git\n",
    "cd padloc-db/sys\n",
    "rm *other* cas_adaptation.yaml\n",
    "cd ..\n",
    "cp sys_meta.txt sys_meta_original.txt\n",
    "grep -v other sys_meta_original.txt | grep -v cas_adaptation > sys_meta.txt\n",
    "cd ..\n",
    "```\n",
    "\n",
    "now we can launch the prediction\n",
    "\n",
    "```shell\n",
    "cd DATA/P.aeruginosa/PADLOC\n",
    "mkdir Analysis\n",
    "for gff in tmp_genomes/GFF/*; do org=$(echo $gff | cut -d'/' -f1,2,5 | sed 's/_genomic.gff//'); faa=$(echo $gff | sed 's/GFF/FAA/' |sed 's/_genomic.gff/_protein.faa/'); mkdir Analysis/$org; echo padloc --gff $gff --faa $faa --cpu 1 -o Analysis/$org --data ../../padloc-db; done\n",
    "```\n",
    "\n",
    "(Optional) You can delete the uncompressed genome files, such as:\n",
    "```shell\n",
    "rm -rf tmp_genomes/*\n",
    "```\n",
    "\n",
    "### Prediction with Defense Finder\n",
    "First, we need to uncompress the translated CDS in order to have similar sequences between tools.\n",
    "\n",
    "```shell\n",
    "cd DATA/P.aeruginosa\n",
    "mkdir -p DFinder/tmp_genomes/TransCDS\n",
    "for faa in Genomes/*/files/*translated*;do name=$(basename \"$faa\" .gz); gunzip -c $faa | awk '/^>/ {if (skip) skip=0; if (/pseudo=true/) skip=1} !skip' > DFinder/tmp_genomes/TransCDS/$name; done\n",
    "```\n",
    "\n",
    "DefenseFinder introduce canonical model for CRISPRCas systems. In our analyses we removed these models.\n",
    "\n",
    "```shell\n",
    "cd DATA\n",
    "mkdir DFinderDB\n",
    "cp -r ~/.macsyfinder/models/CasFinder DFinderDB/.\n",
    "cp -r ~/.macsyfinder/models/defense-finder-models DFinderDB/\n",
    "rm DFinderDB/CasFinder/definitions/CAS_Cluster.xml\n",
    "```\n",
    "\n",
    "now we can launch the prediction\n",
    "\n",
    "```shell\n",
    "conda activate defensefinder\n",
    "cd DATA/P.aeruginosa/DFinder/\n",
    "mkdir Analysis\n",
    "for faa in tmp_genomes/TransCDS/*.faa;do name=$(echo $faa | cut -d'/' -f7 | sed 's/_translated_cds.faa//'); mkdir Analysis/$name; echo defense-finder run -o $name -w 1 $faa --models-dir ../../DFinderDB ; done > defense_finder.cmd\n",
    "```\n",
    "\n",
    "(Optional) You can delete the uncompressed genome files, such as:\n",
    "```shell\n",
    "rm -rf P.aeruginosa/DFinder/tmp_genomes/*\n",
    "```\n",
    "\n",
    "### Prediction with PANORAMA\n",
    "\n",
    "First, you need to download the models.\n",
    "\n",
    "<!--Partie à réaliser lorsque les modèles seront mis à disposition-->\n",
    "\n",
    "Now you can predict the systems with PANORAMA\n",
    "\n",
    "```shell\n",
    "conda activate panorama\n",
    "cd DATA/P.aeruginosa\n",
    "mkdir -p PANORAMA/Analysis\n",
    "cd PANORAMA/Analysis\n",
    "echo \"P.aeruginosa\t../../Pangenome/pangenome.h5\" > pangenomes.tsv\n",
    "panorama pansystems -p pangenomes.tsv -s padloc --hmm ../../../Models/padloc/hmm_list.tsv --models ../../../Models/padloc/models_list.tsv -o . --projection --association all --threads 1 --force\n",
    "\n",
    "panorama pansystems -p pangenomes.tsv -s defensefinder --hmm ../../../Models/defense-finder/hmm_list.tsv --models ../../../Models/defense-finder/models_list.tsv -o . --projection --association all --threads 1 --force\n",
    "```\n",
    "\n",
    "## Enterobacteriaceae\n",
    "\n",
    "### Download Genomes\n",
    "\n",
    "```shell\n",
    "conda activate panorama\n",
    "cd DATA/Enterobacteriaceae\n",
    "mkdir Genomes\n",
    "mkdir Genomes/C.freundii\n",
    "genome_updater.sh -M gtdb -e assembly_summary_c_freundii.txt -f 'genomic.gbff.gz' -o Genomes/C.freundii -t 1\n",
    "mkdir Genomes/E.coli\n",
    "genome_updater.sh -M gtdb -e assembly_summary_e_coli.txt -f 'genomic.gbff.gz' -o Genomes/E.coli -t 1\n",
    "mkdir Genomes/K.pneumoniae\n",
    "genome_updater.sh -M gtdb -e assembly_summary_k_pneumoniae.txt -f 'genomic.gbff.gz' -o Genomes/K.pneumoniae -t 1\n",
    "mkdir Genomes/S.enterica\n",
    "genome_updater.sh -M gtdb -e assembly_summary_s_enterica.txt -f 'genomic.gbff.gz' -o Genomes/S_enterica -t 1\n",
    "```\n",
    "### Generate the pangenomes\n",
    "\n",
    "First we generate the input files for PPanGGOLiN.\n",
    "\n",
    "```shell\n",
    "for gbff in Genomes/C.Freundii/2024-07-09_19-21-03/files/*gbff*; do org=$(echo $gbff | cut -d'/' -f5 | sed 's/_genomic.gbff.gz//'); echo $org $(pwd)/$gbff; done > Pangenomes/C.Freundii_gbff.list; sed -i 's/\\s\\s*/\\t/' Pangenomes/C.Freundii_gbff.list\n",
    "for gbff in Genomes/E.Coli/2024-07-09_19-20-03/files/*gbff*; do org=$(echo $gbff | cut -d'/' -f5 | sed 's/_genomic.gbff.gz//'); echo $org      $(pwd)/$gbff; done > Pangenomes/E.coli_gbff.list; sed -i 's/\\s\\s*/\\t/' Pangenomes/E.coli_gbff.list\n",
    "for gbff in Genomes/K.pneumoniae/2024-07-09_19-22-03/files/*gbff*; do org=$(echo $gbff | cut -d'/' -f5 | sed 's/_genomic.gbff.gz//'); echo $org       $(pwd)/$gbff; done > Pangenomes/K.pneumoniae_gbff.list; sed -i 's/\\s\\s*/\\t/' Pangenomes/K.pneumoniae_gbff.list\n",
    "for gbff in Genomes/S.Enterica/2024-07-09_19-24-34/files/*gbff*; do org=$(echo $gbff | cut -d'/' -f5 | sed 's/_genomic.gbff.gz//'); echo $org $(pwd)/$gbff; done > Pangenomes/S.Enterica_gbff.list; sed -i 's/\\s\\s*/\\t/' Pangenomes/S.Enterica_gbff.list\n",
    "```\n",
    "\n",
    "We can now generate pangenomes\n",
    "```shell\n",
    "cd Pangenomes\n",
    "ppanggolin all --anno C.Freundii_gbff.list --no_flat_files --cpu 36 -o C.freundii -f\n",
    "ppanggolin all --anno E.coli_gbff.list --no_flat_files --cpu 1 -o E.coli -f\n",
    "ppanggolin all --anno K.pneumoniae_gbff.list --no_flat_files --cpu 1 -o K.pneumoniae -f\n",
    "ppanggolin all --anno S.Enterica_gbff.list --no_flat_files --cpu 1 -o S.Enterica -f\n",
    "cd ..\n",
    "```\n",
    "\n",
    "### Run the analysis with PANORAMA\n",
    "\n",
    "We will run the analysis with the translated defense finder models, previously used for *P.aeruginosa*\n",
    "\n",
    "```shell\n",
    "mkdir Analyses\n",
    "cd Analyses\n",
    "echo \"C.freundii\tPangenomes/C.freundii/pangenome.h5\" > pangenomes.tsv\n",
    "echo \"E.coli\tPangenomes/E.coli/pangenome.h5\" >> pangenomes.tsv\n",
    "echo \"K.pneumoniae\tPangenomes/K.pneumoniae/pangenome.h5\" >> pangenomes.tsv\n",
    "echo \"S.enterica\tPangenomes/S.enterica/pangenome.h5\" >> pangenomes.tsv\n",
    "panorama pansystems -p pangenomes.tsv -s dfinder --hmm ../../Models/defense-finder/hmm_list.tsv --models ../../Models/defense-finder/models_list.tsv -o . --projection --association all --threads 1 --force\n",
    "```\n",
    "\n",
    "## Run the Jupyter Notebook\n",
    "\n",
    "You now have all the results. You can rerun the notebook to get the figures."
   ],
   "id": "916894ba09e47314"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "",
   "id": "50fe68cd91f76d9a",
   "outputs": [],
   "execution_count": null
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "panorama_notebook",
   "language": "python",
   "name": "panorama_notebook"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
